src/main/scala/chisel3/util/Mux.scala

```diff
@@ -49,21 +52,47 @@ object PriorityMux {
   def apply[T <: Data](sel: Bits, in: Seq[T]): T = apply((0 until in.size).map(sel(_)), in)
 }
 
-/** Creates a cascade of n Muxs to search for a key value.
+/** Creates a cascade of n Muxs to search for a key value. The Selector may be a UInt or an EnumType.
   *
   * @example {{{
-  * MuxLookup(idx, default,
-  *     Array(0.U -> a, 1.U -> b))
+  * MuxLookup(idx, default)(Seq(0.U -> a, 1.U -> b))
+  * MuxLookup(myEnum, default)(Seq(MyEnum.a -> 1.U, MyEnum.b -> 2.U, MyEnum.c -> 3.U))
   * }}}
   */
 object MuxLookup {
 
+  /** Creates a cascade of n Muxs to search for a key value.
+    *
+    * @example {{{
+    * MuxLookup(idx, default, Seq(0.U -> a, 1.U -> b))
+    * }}}
+    *
+    * @param key a key to search for
+    * @param default a default value if nothing is found
+    * @param mapping a sequence to search of keys and values
+    * @return the value found or the default if not
+    */
+  @deprecated("Use MuxLookup(key, default)(mapping) instead", "Chisel 3.6")
+  def apply[S <: UInt, T <: Data](key: S, default: T, mapping: Seq[(S, T)]): T =
+    macro MuxLookupTransform.apply[S, T]
+
+  /** @param key a key to search for
+    * @param default a default value if nothing is found
+    * @param mapping a sequence to search of keys and values
+    * @return the value found or the default if not
+    */
+  def apply[T <: Data](key: UInt, default: T)(mapping: Seq[(UInt, T)]): T =
+    macro MuxLookupTransform.applyCurried[UInt, T]
+
   /** @param key a key to search for
     * @param default a default value if nothing is found
     * @param mapping a sequence to search of keys and values
     * @return the value found or the default if not
     */
-  def apply[S <: UInt, T <: Data](key: S, default: T, mapping: Seq[(S, T)]): T = {
+  def apply[T <: Data](key: EnumType, default: T)(mapping: Seq[(EnumType, T)]): T =
```

Comment:
Shouldn't this still have some type E that extends an enumtype? Like I should not be able to mix days of the week with months

Reply:
ah yeah that's still a useful constraint to have